src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceTypeCompleter.cs

```diff
@@ -0,0 +1,149 @@
+//
+// Copyright Microsoft Corporation
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+namespace Microsoft.Azure.Commands.ResourceManager.Common.ArgumentCompleters
+{
+    using Commands.Common.Authentication;
+    using Commands.Common.Authentication.Abstractions;
+    using Management.Internal.Resources;
+    using Management.Internal.Resources.Models;
+    using Properties;
+    using Rest.Azure;
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Management.Automation;
+
+    /// <summary>
+    /// This attribute will allow the user to autocomplete the -ResourceType parameter of a cmdlet with valid resource types
+    /// </summary>
+    public class ResourceTypeCompleterAttribute : ArgumentCompleterAttribute
+    {
+        private static IDictionary<int, IList<String>> _resourceTypesDictionary = new ConcurrentDictionary<int, IList<string>>();
+        private static readonly object _lock = new object();
+        public static int _timeout = 3;
+
+        protected static IList<String> ResourceTypes
+        {
+            get
+            {
+                lock (_lock)
+                {
+                    IAzureContext context = AzureRmProfileProvider.Instance.Profile.DefaultContext;
+                    var contextHash = HashContext(context);
+                    if (!_resourceTypesDictionary.ContainsKey(contextHash))
+                    {
+                        var tempResourceTypeList = new List<string>();
+                        try
+                        {
+                            var instance = AzureSession.Instance;
+                            var client = instance.ClientFactory.CreateCustomArmClient<ResourceManagementClient>(
+                                context.Environment.GetEndpointAsUri(AzureEnvironment.Endpoint.ResourceManager),
+                                instance.AuthenticationFactory.GetServiceClientCredentials(context, AzureEnvironment.Endpoint.ResourceManager),
+                                instance.ClientFactory.GetCustomHandlers());
+                            client.SubscriptionId = context.Subscription.Id;
+                            // Retrieve only the first page of resource types to display to the user
+                            var resourceTypes = client.Providers.ListAsync();
```

Comment:
why only one page?

Reply:
@markcowl looks like this was a copy-paste issue -- I changed this so it retrieves all pages